astropy/io/fits/file.py

```diff
@@ -119,10 +121,16 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
         if mode not in IO_FITS_MODES:
             raise ValueError("Mode '{}' not recognized".format(mode))
 
+        # Handle raw URLs
         if (isinstance(fileobj, string_types) and
-            mode not in ('ostream', 'append') and
-            _is_url(fileobj)):  # This is an URL.
+            mode not in ('ostream', 'append', 'update') and _is_url(fileobj)):
             self.name = download_file(fileobj, cache=cache)
+        # Handle responses from URL requests that have already been opened
+        elif isinstance(fileobj, http_client.HTTPResponse):
+            if mode in ('ostream', 'append', 'update'):
+                raise ValueError(
+                    "Mode {} not supported for HTTPResponse".format(mode))
+            fileobj = io.BytesIO(fileobj.read())
```

Comment:
If you pass a Response object, I think you have to expect that it will be used to read the data. Otherwise there are other ways, passing the url or saving the data to disk.